database/raft/raft.go

```diff
@@ -0,0 +1,920 @@
+	defer sv.stateMu.Unlock()
+	for !sv.done && sv.state.AppliedIndex() < index {
+		sv.stateCond.Wait()
+	}
+	return sv.state.AppliedIndex() >= index //if false killed b/c of done signal
+}
+
+// Stale returns an object that reads
+// directly from local memory, returning (possibly) stale data.
+// Calls to sv.Get are linearizable,
+// which requires them to go through the raft protocol.
+// The stale getter skips this, so it is much faster,
+// but it can only be used in situations that don't require
+// linearizability.
+func (sv *Service) Stale() Getter {
+	return (*staleGetter)(sv)
+}
+
+// ServeHTTP responds to raft consensus messages at /raft/x,
+// where x is any particular raft internode RPC.
+// When sv sends outgoing messages, it acts as an HTTP client
+// and sends requests to its peers at /raft/x.
+func (sv *Service) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	sv.mux.ServeHTTP(w, req)
+}
+
+func (sv *Service) serveMsg(w http.ResponseWriter, req *http.Request) {
+	b, err := ioutil.ReadAll(http.MaxBytesReader(w, req.Body, maxRaftReqSize))
+	if err != nil {
+		http.Error(w, "cannot read req: "+err.Error(), 400)
+		return
+	}
+	var m raftpb.Message
+	err = m.Unmarshal(b)
+	if err != nil {
+		http.Error(w, "cannot unmarshal: "+err.Error(), 400)
+		return
+	}
+	sv.raftNode.Step(req.Context(), m)
+}
+
+func (sv *Service) serveJoin(w http.ResponseWriter, req *http.Request) {
+	var x struct{ Addr string }
+	err := json.NewDecoder(req.Body).Decode(&x)
+	if err != nil {
+		http.Error(w, err.Error(), 400)
+		return
+	}
+
+	log.Write(req.Context(), "at", "join", "addr", x.Addr)
+
+	newID, err := sv.allocNodeID(req.Context())
+	if err != nil {
+		http.Error(w, err.Error(), 400)
+		return
+	}
+
+	log.Write(req.Context(), "at", "join-id", "addr", x.Addr, "id", newID)
+
+	err = sv.raftNode.ProposeConfChange(req.Context(), raftpb.ConfChange{
+		ID:      atomic.AddUint64(&sv.confChangeID, 1),
+		Type:    raftpb.ConfChangeAddNode,
+		NodeID:  newID,
+		Context: []byte(x.Addr),
+	})
+	if err != nil {
+		http.Error(w, err.Error(), 500)
+		return
+	}
+
+	log.Write(req.Context(), "at", "join-done", "addr", x.Addr, "id", newID)
+
+	snap := sv.getSnapshot()
+	snapData, err := encodeSnapshot(snap)
+	if err != nil {
+		http.Error(w, err.Error(), 500)
+		return
+	}
+	json.NewEncoder(w).Encode(struct {
+		ID   uint64
+		Snap []byte
+	}{newID, snapData})
+}
+
+// join attempts to join the cluster.
+// It requests an existing member to propose a configuration change
+// adding the local process as a new member, then retrieves its new ID
+// and a snapshot of the cluster state and applies it to sv.
+// It also sets sv.id.
+func (sv *Service) join(addr, baseURL string) error {
+	reqURL := strings.TrimRight(baseURL, "/") + "/raft/join"
+	b, _ := json.Marshal(struct{ Addr string }{addr})
+	resp, err := http.Post(reqURL, contentType, bytes.NewReader(b))
+	if err != nil {
+		return errors.Wrap(err)
+	}
+	defer resp.Body.Close()
+
+	var x struct {
```

Comment:
Might want to factor this out into a named type since we also use this same type in serveJoin.